### PR TITLE
Collection expressions: require [CollectionBuilder] target iteration type from GetEnumerator instance method

### PR DESCRIPTION
**Pull request by cston**
See [LDM-2024-01-08](https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-01-08.md#conclusion).

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 1]
```diff
@@ -785,14 +785,15 @@ BoundNode bindSpreadElement(BoundCollectionExpressionSpreadElement element, Type
             return element;
         }
 
-        internal bool TryGetCollectionIterationType(ExpressionSyntax syntax, TypeSymbol collectionType, out TypeWithAnnotations iterationType)
+        internal bool TryGetCollectionExpressionIterationType(ExpressionSyntax syntax, TypeSymbol collectionType, out TypeWithAnnotations iterationType)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:788]: >TryGetCollectionExpressionIterationType [](http://example.com/codeflow?start=22&length=39)

Consider leaving the name as is. This isn't really about an expression, this is about a type. For example, this method is used by "Params Collections" feature. There isn't even a collection expression in the picture for those scenarios. Another thing to consider, is the fact that I changed the signature of this method in [https://github.com/dotnet/roslyn/tree/features/ParamsCollections](https://github.com/dotnet/roslyn/tree/features/ParamsCollections) to use SyntaxNode instead of ExpressionSyntax due to the reason outlined above. The rename is likely to cause conflicts during integrations, which I would prefer to avoid.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 2]
```diff
@@ -785,14 +785,15 @@ BoundNode bindSpreadElement(BoundCollectionExpressionSpreadElement element, Type
...
         {
             BoundExpression collectionExpr = new BoundValuePlaceholder(syntax, collectionType);
             return GetEnumeratorInfoAndInferCollectionElementType(
                 syntax,
                 syntax,
                 ref collectionExpr,
                 isAsync: false,
+                allowExtensionMethod: false, // collection expression target types do not require extension method GetEnumerator
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:796]: >allowExtensionMethod: false, // collection expression target types do not require extension method GetEnumerator [](http://example.com/codeflow?start=16&length=112)

I suggest to not add this parameter. It looks like we can simply check ```builder.ViaExtensionMethod``` when `GetEnumeratorInfoAndInferCollectionElementType` succeeds.

**Comment by AlekseyTs** [comment]
I think we should also address #71172.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

**Comment by cston** [comment]
> I think we should also address #71172.

Yes, I am planning on fixing #71172 (by validating the attribute in `SourceNamedTypeSymbol.DecodeWellKnownAttributeImpl`) but I'd prefer to do that in a follow-up PR to reduce the risk here (from binding cycles for instance) in case we want to take this change in 17.9.

File: src/Compilers/CSharp/Test/Emit/CodeGen/ForeachTest.cs [hunk 3]
```diff
@@ -1653,5 +1653,43 @@ static void Main()
...
+                    {
+                        MyCollection<int> x = new(1, 2, 3);
+                        int total = 0;
+                        foreach (var y in x)
+                            total += y;
+                    }
+                }
+                """;
+
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularDefault.WithFeature("run-nullable-analysis", "never"));
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/ForeachTest.cs:1683]: >.WithFeature("run-nullable-analysis", "never") [](http://example.com/codeflow?start=89&length=46)

Why is this needed?

**Comment by AlekseyTs** [comment]
> I'd prefer to do that in a follow-up PR to reduce the risk here (from binding cycles for instance) in case we want to take this change in 17.9.

Sounds reasonable.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 4]
```diff
@@ -13519,6 +13519,54 @@ static void Main()
                 Diagnostic(ErrorCode.ERR_CollectionBuilderAttributeMethodNotFound, "[null]").WithArguments("Create", "T", "MyCollection<T>").WithLocation(7, 34));
         }
 
+        [Fact]
+        public void CollectionBuilder_MissingBuilderMethod_FromMetadata()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:13523]: Discussed offline: Something is strange with this test

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 3)

File: src/Compilers/CSharp/Test/Emit/CodeGen/ForeachTest.cs (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Emit/CodeGen/ForeachTest.cs:1683]: This is needed to ensure we don't set `ForEachLocalSymbol._type` ahead of the call to `InferTypeOfVarVariable` which we're trying to test here. That method calls `ForEachLoopBinder.InferCollectionElementType` which should allow extension method `GetEnumerator`.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5), assuming CI is passing

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 5)
